libevmasm/Assembly.cpp

```diff
@@ -539,14 +539,36 @@ LinkerObject const& Assembly::assemble() const
 	unsigned bytesPerTag = numberEncodingSize(bytesRequiredForCode);
 	uint8_t tagPush = static_cast<uint8_t>(pushInstruction(bytesPerTag));
 
-	unsigned bytesRequiredIncludingData = bytesRequiredForCode + 1 + static_cast<unsigned>(m_auxiliaryData.size());
+	unsigned bytesRequiredIncludingData = (eof ? 10 : 0) + bytesRequiredForCode + 1 + static_cast<unsigned>(m_auxiliaryData.size());
 	for (auto const& sub: m_subs)
 		bytesRequiredIncludingData += static_cast<unsigned>(sub->assemble().bytecode.size());
 
 	unsigned bytesPerDataRef = numberEncodingSize(bytesRequiredIncludingData);
 	uint8_t dataRefPush = static_cast<uint8_t>(pushInstruction(bytesPerDataRef));
 	ret.bytecode.reserve(bytesRequiredIncludingData);
 
+	// Insert EOF1 header.
+	bytesRef eofCodeLength(&ret.bytecode.back(), 0);
+	bytesRef eofDataLength(&ret.bytecode.back(), 0);
+	if (eof)
+	{
+		// TODO: empty data is disallowed
+		ret.bytecode.push_back(0xef);
+		ret.bytecode.push_back(0x00);
+		ret.bytecode.push_back(0x01); // version 1
+		ret.bytecode.push_back(0x01); // kind=code
```

Comment:
This header is fine for single-code section case. For functions it needs to be generalised and needs a type section too.